contracts/DutchAuction.sol

```diff
@@ -0,0 +1,176 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.4;
+
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import "@openzeppelin/contracts/access/AccessControl.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/token/common/ERC2981.sol";
+import "erc721a/contracts/ERC721A.sol";
+
+import "./IDutchAuction.sol";
+import "./ERC721M.sol";
+
+contract DutchAuction is IDutchAuction, ERC721M, ReentrancyGuard {
+    bool private immutable _refundable;
+    uint256 private _settledPriceInWei;
+    Config private _config;
+    mapping(address => User) private _userData;
+
+    constructor(
+        // ERC721M
+        string memory collectionName,
+        string memory collectionSymbol,
+        string memory tokenURISuffix,
+        uint256 maxMintableSupply,
+        uint256 globalWalletLimit,
+        address cosigner,
+        bool refundable
+    )
+        ERC721M(
+            collectionName,
+            collectionSymbol,
+            tokenURISuffix,
+            maxMintableSupply,
+            globalWalletLimit,
+            cosigner
+        )
+    {
+        _refundable = refundable;
+    }
+
+    modifier validTime() {
+        Config memory config = _config;
+        if (
+            block.timestamp > config.endTime ||
+            block.timestamp < config.startTime
+        ) revert InvalidStartEndTime();
+        _;
+    }
+
+    function setConfig(
+        uint256 startAmountInWei,
+        uint256 endAmountInWei,
+        uint256 startTime,
+        uint256 endTime,
+        bool roundUp
+    ) external onlyOwner {
+        if (startTime == 0 || startTime >= endTime)
+            revert InvalidStartEndTime();
```

Comment:
nit: can include the `startTime` and `endTime` as parameters in the error https://blog.soliditylang.org/2021/04/21/custom-errors/